msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1064,51 +1064,197 @@ protected void acquireTokenSilent(@NonNull final String[] scopes,
+        // The root account we'll be fetching tokens for...
+        final IAccount rootAccount = tokenParameters.getAccount();
+        final MultiTenantAccount multiTenantAccount = (MultiTenantAccount) rootAccount;
+        final String requestAuthority = tokenParameters.getAuthority();
+        final Authority authority = Authority.getAuthorityFromAuthorityUrl(requestAuthority);
 
-        try {
-            final TokenCommand silentTokenCommand = new TokenCommand(
-                    params,
-                    MSALControllerFactory.getAcquireTokenSilentControllers(
-                            mPublicClientConfiguration.getAppContext(),
-                            params.getAuthority(),
-                            mPublicClientConfiguration
-                    ),
-                    callback
+        if (authority instanceof AzureActiveDirectoryB2CAuthority) {
+            // use home account - b2c is not compatible with broker, so no need to construct
+            // the account used in the request...
+            return AccountAdapter.getAccountInternal(
+                    mPublicClientConfiguration.getClientId(),
+                    mPublicClientConfiguration.getOAuth2TokenCache(),
+                    multiTenantAccount.getHomeAccountId(),
+                    multiTenantAccount.getTenantId()
+            );
+        } else if (authority instanceof AzureActiveDirectoryAuthority) {
+            final AzureActiveDirectoryAuthority aadAuthority = (AzureActiveDirectoryAuthority) authority;
+
+            // Although the below call implies the returned value will be a tenantId, this isn't
+            // strictly true: it can also be an alias such as 'common', 'consumers', or
+            // 'organizations'. Additionally, if the developer has used a named tenant it could end
+            // up being something like <tenant_name>.onmicrosoft.com
+            //
+            // If the tenant is a GUID, we need to choose the account/profile that corresponds to it
+            //
+            // If the tenant is named 'common', 'consumers', or 'organizations' we need to use the
+            // home account
+            //
+            // If the account is named like <tenant_name>.onmicrosoft.com we need to query the OpenId
+            // Provider Configuration Metadata in order to get the tenant id. Once we have the
+            // tenant id, we must then select the appropriate home or profile.
+            final String tenantIdNameOrAlias = aadAuthority.getAudience().getTenantId();
+
+            // The AccountRecord we'll use to request a token...
+            final AccountRecord accountRecord = new AccountRecord();
+            accountRecord.setEnvironment(multiTenantAccount.getEnvironment());
+            accountRecord.setHomeAccountId(multiTenantAccount.getHomeAccountId());
+
+            final boolean isUuid = isUuid(tenantIdNameOrAlias);
+
+            if (isUuid) {
```

Comment:
I was actually thinking of something like this
if (!isUuid){
final OpenIdProviderConfiguration providerConfiguration =
                         loadOpenIdProviderConfigurationMetadata(requestAuthority);

                  final String issuer = providerConfiguration.getIssuer();
                 final Uri issuerUri = Uri.parse(issuer);
                 final List<String> paths = issuerUri.getPathSegments();
                 tenantIdNameOrAlias = paths.get(0);
}

then proceed with the current if else.

Reply:
I gotcha - this is cleaner. I'll make that change

Reply:
Resolved in a7e162282275bf8fea08721516ccc29e9671fe1c